### DISCUSSION TITLE
Block quote syntax

### DISCUSSION THREAD

**Discussion by tingerrr** [typst/typst#2361] [Ideas]
As discussed in #2252, I would like to propose adding syntax for blockquotes, i.e. (`quote(block: true)[...]`), I was writing an issue when I realized that it was becoming very long, so I want to get some design questions out of the way first.

# Syntax
markdown provides the following syntax for block quotes:
```
> Hello
> World
```

I think this is a good fit for typst as leading `>` are not yet used for anything. Markdown is also common enough that that users will feel familiar with this syntax.

## Implementation 1
- Add `quote.line`
- Add `quote.attribution`

The given example should yield a CST containing two `QuoteLine` nodes.

The `QuoteLine` syntax nodes would behave the same as `ListItem` currently does, they can be continued using indentation in case of explicit new lines. The following quote line is a single element:
```
> Hello
  World
```

Unlike `ListItem`, there is no merging of individual list items if they are separated by whitespace other than a newline, i.e. the following document produces two individual quotes:
```
> Hello

> World
```

Similar to `ListItem`, each line is its own element and can be individually styled, but would by default simply be joined. In #105 (which or some reason doesn't exist anymore), it was mentioned that the `\quote` macro family in LaTeX is also used to typeset poetry with very specific styling for line indentation. Line breaks may also have significance for normal quotes and would not be preserved without at least an individual line syntax node. This is inspired by #2341 and should likely also carry similar information about the line numbers to allow the poetry styling.

An argument against this, is that line breaks can be inserted explicitly using `\`, but the poetry styling would then have to inspect the sequences.

### Attributions
Attributions could be supported by placing them on a single line at the end, which has a special marker:
```
> ...
>: attribution
```

This would produce a `QuoteAttribution` node and would only be valid if there is at least one `QuoteLine` before it.

The `>:` maker is chosen to avoid bikeshedding for now, the discussion on the exact syntax should be done once the right implementation was chosen.

## Implementation 2
Instead of using the familiar syntax of one `>` for each line, a quote could simply be a single continued line.
```
> Hello
  World
```

### Attributions
Attributions in this syntax could be supported as optional or required prefix/suffix using a separator.
- `> attribution: quote`
  - colons must be escaped in the attribution, rarely an issue
  - empty attribution (`> : quote`) would be interpreted as `none`
- `> quote : attribution`
  - colons must be escaped in the quote, possibly cumbersome
  - no colon is needed if there is no attribution (`> quote`)
  - might introduce unnecessary look ahead/look behind overhead

Both versions use term-like syntax, this is somewhat familiar and makes sense, especially for the first variant: `> René Descartes: cogito, ergo sum` reads like "René Descartes said «cogito ergo sum»".

# Semantics of Attributions
`quote` allows passing a label to unambiguously refer to a bibliography entry, content (and therefore also strings) are interpreted as content to literally insert as the attribution. When a label is passed, it will (in the future, see CSL support) display the author in any case and then attach the citation.

These pose a bit of a problem, if they are supported they could work like this, assuming syntax one of Implementation 2:
- `> <descartes>: cogito, ergo sum`
  - consistent with the code version (`quote(attribution: <descartes>)`)
  - inconsistent with the intuition of placing a label in markdown mode (it behaves like `@descartes`)
- `> @descartes: cogito, ergo sum`
  - consistent with the intuition of placing a label in markdown mode
  - inconsistent with the code version
  - must be special cased to extract the label instead of placing the citation as is
  - if special cased to mean `attribution: <descartes>`, would be ambiguous with `attribution: [@descartes]`

Once again, `:` was chosen as its familiar syntax from term lists, but a different separator could be chosen. Theoretically, `> attribution "quote"` could also be used in case there is an attribution, while keeping `> quote` otherwise.

If label attributions are not special cased as above, then this would simply wrap them in content and no longer produce the desired attribution. `quote(attribution: <descartes>)` is not the same as `quote(attribution: [@descartes])`, only similar for some styles, which include the author.

# Request for Comments
What are your thoughts on this, should this be supported? Should one of these implementations be chosen, or do you have better ideas? How should label attributions be supported in a way that is consistent with the intuitions across code and markup mode.

**Comment by emilyyyylime** [comment]
I'll mention that `> quote : attribution` can simply be 'left-associative', and thus still only require escaping the delimiter in the attribution.
I'm not the biggest fan of `:` as an attribution separator (for either syntax) from an intuition standpoint, though I can see how it makes sense and I could live with it.

Regarding the question of which implementation to use, it seems to me that the latter is strictly a subset of the former — so I'm leaning towards the former. The one thing I can think of that you can't achieve with it is a bunch of quotes put together compactly as in
```
> "Quote 1" -- attribution
> "Quote 2" -- attribution
```
(If we're going with the latter, I do quite like the look of `> "quote" -- attribution`)

**Reply by tingerrr** [reply]
> I'll mention that > quote : attribution can simply be 'left-associative', and thus still only require escaping the delimiter in the attribution.

I think the left-associativity comes with more complicated parsing. I have played around a bit and implemented the reverse case (`> attribution: quote`). And by the looks of it, the forward case would require looking back from the end of the quote in either the lexer or parsing stage.

> Regarding the question of which implementation to use, it seems to me that the latter is strictly a subset of the former — so I'm leaning towards the former. The one thing I can think of that you can't achieve with it is a bunch of quotes put together compactly as in

This was mainly because on #105 (and then subsequently not part of the discussion on #2252), it was noted by @laurmaedje that having sub elements for both source and author (which later became attribution) were already too many sub elements for him. Implementation 1 would introduce 2 sub elements, although for different reasons.

While `> "Quote 1" -- attribution` closely resembles what the quote looks like by default, which is always a plus, I think it requires a few too many tokens for me to be called syntax-sugar. But it's still prettier than `quote(block: true, attribution: [...])[...]`.

I'm personally also leaning more towards Implementation 1.

**Reply by sitandr** [reply]
I don't really like `> "Quote" -- attribution` thing. I don't know much about other languages, but in Russian we use em dashes a lot, especially in quotes (basically to describe some projects properties, replacing "is", "then", "that's why" or similar things). And requiring quotation mark may again interfere with inner quotations (and generally strongly resembles string to me).

I'm afraid that this syntax may conflict with the quote itself, making it ambiguous. Maybe we could require new line or something like that, but I guess it will make it extremely verbose and complex to parse.

**Comment by laurmaedje** [comment]
Thanks for the well-written proposal! One thing I'm missing in it is motivation. I feel like multiple things are at play here:

- In Markdown, _everything_ needs to be syntax. In Typst, we can always have a function, so only the most often used things need syntax.
- Markdown is used often used in forums and conversations where quoting is very common. Meanwhile, Typst's main use case is writing documents, where block quotes are needed far less I think. It's not too bad to just write `#quote[...]`.
- It's very hard to have good syntax for attributions. Since none of the proposed syntaxes for attribution have any precedent (afaik), this would likely be a piece of syntax very few people are aware of. 

Overall, we already have a _lot_ of syntax and adding new syntax has a complexity high cost in my opinion. I specifically like about Typst in comparison to Markdown or AsciiDoc (which has tons of syntax), that we _don't_ need syntax for everything because we have a more general mechanism.

**Reply by tingerrr** [reply]
> - In Markdown, everything needs to be syntax. In Typst, we can always have a function, so only the most often used things need syntax.
> - Markdown is used often used in forums and conversations where quoting is very common. Meanwhile, Typst's main use case is writing documents, where block quotes are needed far less I think. It's not too bad to just write #quote[...].

I wrote quite a few of them while testing the element, and it felt a bit cumbersome, similar to how `link("...")[...]` feels cumbersome. Currently, `quote` has these benefits to the user:
- style label attributions specially depending on the citation style in use
- retrieve the author to display, even if it's not an author style (this is blocked by CSL support but planned)
- allow show and set rules for styling

All three of those points are things that only benefit from being built-in now, as contexts, CSL support and custom elements would allow any user to recreate this in user space. This is not a bad thing, but it begs whether the element should exist by then.

I think, authors of documents focusing on literary analysis could really benefit from this, quoting poetry or otherwise stylized texts would become much more manageable.

Once HTML output becomes available, this syntax could have more demand in general, so perhaps it could be reevaluated at that time.

> - It's very hard to have good syntax for attributions. Since none of the proposed syntaxes for attribution have any precedent (afaik), this would likely be a piece of syntax very few people are aware of.

That's true, I'm not really satisfied with any solution yet, they all come with various caveats even if we ignore the label attribution problem.

But I also think this shouldn't keep us from trying to innovate and come up with a syntax that makes sense.

I'd like to play with the whole idea a bit more, even if it doesn't come to fruition.

**Reply by sitandr** [reply]
I very like the "poetry" part of that idea. Currently there are two basic ways to type things, preserving linebreaks. 

First, add linebreaks after each line, which is really annoying. Other uses `raw`, but you can't use any content there.

I think it would be very cool to have some basic language construct that preserves linebreaks and allows everything inside of it. I'm sure there will be more uses of that, not just poetry.

Maybe it should be something else, not quote, but using quote seems okay to me.

**Reply by tingerrr** [reply]
> Maybe it should be something else, not quote, but using quote seems okay to me.

This mainly came out of a suggestion in #105 (you'll have to believe me, I guess), that this is how people used it in LaTeX because it preserved line breaks the way it was described here.

Then, perhaps, quote is instead a special case of this construct.

> I think it would be very cool to have some basic language construct that preserves linebreaks and allows everything inside of it. I'm sure there will be more uses of that, not just poetry.

If this were an element, then the lack of attribution would also be ok, as it's less part of this specific use case of preserving line breaks.

**Reply by laurmaedje** [reply]
> I think it would be very cool to have some basic language construct that preserves linebreaks and allows everything inside of it. I'm sure there will be more uses of that, not just poetry.

There was also the idea of having a set rule that lets you keep linebreaks. Right now, normal spaces and linebreaks both result in a `SpaceElem` which forgets whether there was a linebreak. But it could easily remember and then we could have some property on the then-exposed `space` element to turn them into `LinebreakElem`s.

**Comment by Dherse** [comment]
Just came here to say the following: I like it 👍

**Comment by CGMossa** [comment]
I would also like to emphasize, that having precise syntax for quotes, will help with reading the text file itself as a document. To be fair, it is an advantage of typst, that the syntax is precise, and without redundancy, take math mode as an example.

I would also say that the attribution stuff could be used for code blocks as well.
Developing a good syntax for this, could be a precursor for attribution of code blocks.

**Comment by tingerrr** [comment]
Closing this as not planned, see #2362 and feel free to reopen this if you have a compelling argument for actually.
